Feature/database storage

emmaLP opened the pull request:
Changes:
*  Adding a database client for communication with a mongo db. Including tests
* Updated the service code to call the database saveItem method 
  * Adding additional tests and updated existing ones

emmaLP committed:
Updating gitignore to ignore any files in the bin/ folder

emmaLP committed:
Building Dockerfile to support multiple docker images been built

emmaLP committed:
Initial docker compose file for mongo db and consumer

emmaLP committed:
Merge branch 'feature/consumer' into feature/database_storage

emmaLP committed:
Merge branch 'main' into feature/database_storage

emmaLP committed:
Changing the test pipeline to only run on commits

emmaLP committed:
Adding a PR pipeline to perform additional checks on a PR

emmaLP committed:
Base structure for the database client

emmaLP committed:
Merge branch 'main' into feature/database_storage

emmaLP committed:
Database functionality and writing unit and integration tests

emmaLP committed:
Fixing the database interface and method implementation for SaveItem method

emmaLP committed:
Updating the service and tests to handle saving a hackernew items to then database

emmaLP committed:
Removing macOs from the matrix as it doesn't come with docker installed by default

emmaLP committed:
Adding go version arg

emmaLP committed:
Removing window os from matrix

emmaLP committed:
Rename Pr job name

emmaLP committed:
whitespace

emmaLP committed:
Adding the base url to the docker-compose and making the go_version arg global

emmaLP committed:
Adding interrupt signal and simplifying dockerfile

emmaLP committed:
Updating readme

emmaLP committed:
Fixing whitespaces between imports

alexdunne commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+COPY ./internal/ ./internal/
+COPY ./pkg/ ./pkg/
+
+RUN go mod download
```
It might be worth moving this to be directly after the copying of `go.mod` and `go.sum` but before the copying of the rest of the code so Docker's caching is utilised and the dependencies are only downloaded again if `go.mod` or `go.sum` change.

emmaLP replied:
Done

alexdunne commented on README.md:
````diff
@@ -13,12 +13,28 @@ The consumer will periodically run to seed a mongo database from the HackerNews
+docker-compose up -d
+```
+
+**_Note: Ctrl-C command on just `docker-compose up` seemed to hang, but running in the background and
````
If the program is hanging on `Ctrl-C` it's likely there's a blocking channel somewhere that's not allowing the program to exit.

emmaLP replied:
Interestingly enough the docker containers are all stopped and in exited state. I'd expect the same blocking in background mode when running docker-compose down as hitting Ctrl-c in tail mode (docker-compose up)

emmaLP replied:
I figured it out, the restart key in mongo was causing the hang. I've removed it as now works with just docker-compose up

alexdunne commented on internal/consumer/service.go:
```diff
@@ -6,37 +6,43 @@ import (
 	numberOfWorkers int
 	hnClient        hackernews.Client
+	dbClient        database.Client
+	context         context.Context
```
It's recommended to pass `context` as an argument rather than storing in on a struct. [See more](https://go.dev/blog/context-and-structs)

emmaLP replied:
Thanks for sending the info over, it was a good read. will get this updated

emmaLP committed:
PR comment improvements

emmaLP committed:
Removing the restart in mongo container as it causes the docker-compose to hang

alexdunne commented on internal/consumer/cron.go:
```diff
@@ -13,7 +14,12 @@ func ConfigureCron(ctx context.Context, logger *zap.Logger, config *model.Config
+		return fmt.Errorf("Unexpected error when connecting to the database. %w", err)
+	}
+	defer databaseClient.CloseConnection()
+	service, err := NewService(logger, config, nil, databaseClient)
```
Just noticed this is passing in `nil` for the optional http client which fallbacks to creating its own version. It might be worth using the Functional Options styles to provide overrides. That way your default api will be nicer (i.e. no `nil`s) but still provides the option to swap out implementations in tests etc.

([See more](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis))

emmaLP replied:
Done, thanks for the info.

emmaLP committed:
Making hackerNews client optional

emmaLP committed:
Naming job as pr file name

alexdunne commented on internal/consumer/service.go:
```diff
@@ -23,19 +23,23 @@ type Client interface {
 
-func NewService(logger *zap.Logger, config *model.Configuration, hnClient hackernews.Client, dbClient database.Client) (*service, error) {
-	if hnClient == nil {
+func NewService(logger *zap.Logger, config *model.Configuration, dbClient database.Client, hnClient ...hackernews.Client) (*service, error) {
```
The Functional Options approach is slightly different than using variadic arguments to allow optional config. [This article](https://www.sohamkamani.com/golang/options-pattern/#defining-the-functional-options) is a bit more succinct than the one I linked originally.

Essentially you define a type that's a function which accepts the type you want to modify e.g.

`type ServiceOption func (*service)`

and then accept a variadic argument of that type:

```
func NewService(logger *zap.Logger, config *model.Configuration, dbClient database.Client, ...opts ServiceOption) (*service, error) {
    s := &service{
        // add defaults here
    }

    // here we use the functional options to override defaults
    for _, opt := range opts {
        opt(s)
    }
}
```

You can then define specific functional options to configure different parts of the service's defaults

```
func WithHttpClient(client http.Client) ServiceOption {
    return func(s *service) {
        s.client = client
    }
}       
```

```
// another example just to demo
func WithLogger(logger *zap.Logger) ServiceOption {
    return func(s *service) {
        s.logger = logger
    }
}       
```

Because the type uses a pointer, these changes will update the original copy during the loop outlined above

You can then use 0-n of these options when creating the service:

```
// default
NewService(logger, config, db)
```

```
// in tests
NewService(logger, config, db, WithHttpClient(client))
```

emmaLP replied:
Done

emmaLP committed:
Implement functional options

JakeHumphries commented on internal/database/client_intergration_test.go:
```diff
@@ -0,0 +1,85 @@
+	require.NoError(t, err)
+	require.NotNil(t, mongo)
+	require.NotNil(t, dbConfig)
+	defer mongo.Terminate(context.TODO())
```
should we check for an error here?

emmaLP replied:
I've recently had this conversation with @alexdunne  around handling errors on defer calls and seem the accepted practice is not to.

Our production code is littered with examples like this one, also this been a test class I'm not too worried if that fails since a user will likely need to fix the test to pass in the first places

JakeHumphries commented on internal/consumer/service_test.go:
```diff
@@ -13,63 +14,92 @@ import (
+					BaseUrl:         "test.com",
+					CronSchedule:    "",
+					NumberOfWorkers: 2,
+				}},
```
NIT: these could be indented:

emmaLP replied:
Fixed

JakeHumphries commented on internal/consumer/service_test.go:
```diff
@@ -13,63 +14,92 @@ import (
 				hnMock.On("GetTopStories").Return([]int{1}, nil)
 				hnMock.On("GetItem", 1).Return(&hnModel.Item{ID: 1}, nil)
+				dbMock.On("SaveItem", &hnModel.Item{ID: 1}).Return(nil)
+
```
NIT: blank line

JakeHumphries commented on internal/consumer/service_test.go:
```diff
@@ -13,63 +14,92 @@ import (
+					BaseUrl:         "test.com",
+					CronSchedule:    "",
+					NumberOfWorkers: 2,
+				}},
```
NIT: these could be indented

emmaLP replied:
fixed

JakeHumphries commented on internal/consumer/service_test.go:
```diff
@@ -13,63 +14,92 @@ import (
+					BaseUrl:         "test.com",
+					CronSchedule:    "",
+					NumberOfWorkers: 2,
+				}},
```
NIT: these could be indented

emmaLP replied:
Fixed

JakeHumphries commented on internal/consumer/service_test.go:
```diff
@@ -13,63 +14,92 @@ import (
+					BaseUrl:         "test.com",
+					CronSchedule:    "",
+					NumberOfWorkers: 2,
+				}},
```
NIT: these could be indented

emmaLP replied:
No sure what you mean by that?

all the formatting is done using go fmt and looks indented to me

JakeHumphries replied:
if you run `gofumpt -w -extra . ` you will see the below...

```
		"One Item": {
			hnMock: &hackernews.Mock{},
			dbMock: &database.Mock{},
			ids:    []int{1},
			config: &model.Configuration{
				Consumer: model.ConsumerConfig{
					BaseUrl:         "test.com",
					CronSchedule:    "",
					NumberOfWorkers: 2,
				},
			},

			expectedMocks: func(t *testing.T, hnMock *hackernews.Mock, dbMock *database.Mock) {
				hnMock.On("GetTopStories").Return([]int{1}, nil)
				hnMock.On("GetItem", 1).Return(&hnModel.Item{ID: 1}, nil)
				dbMock.On("SaveItem", &hnModel.Item{ID: 1}).Return(nil)
			},
		},
```

JakeHumphries commented on internal/consumer/service.go:
```diff
@@ -6,37 +6,53 @@ import (
+type ServiceOptions func(*service)
+
+func NewService(logger *zap.Logger, config *model.Configuration, dbClient database.Client, opts ...ServiceOptions) (*service, error) {
+
```
NIT: blank space

JakeHumphries commented on internal/database/client.go:
```diff
@@ -0,0 +1,79 @@
+		}
+		opts = opts.SetAuth(credentials)
+	}
+	client, err := mongo.Connect(ctx, opts)
```
Might be worth checking to see if the database is ready. you can do something like this... 

```
	for {
		select {
		case <-ctx.Done():
			return nil, fmt.Errorf("mongo refused to connect: %v %w", ctx.Err(), err)
		default:
			err := client.Ping(ctx, readpref.Primary())
			if err == nil {
				log.Print("mongo is now connected")
				return client, nil
			}
		}
	}
```

emmaLP committed:
Waiting for db to ready and removing unit tests that break because the New method relies on a real db connection.

Updating the integration tests to include additional cleanup which adds more coverage

emmaLP committed:
PR comments - minor formatting

emmaLP committed:
Adding some more info to the README

emmaLP committed:
Implementing gofumpt formatting
